### DISCUSSION TITLE
Raylib In Microsoft Small Basic

### DISCUSSION THREAD

**Discussion by zs-3** [raysan5/raylib#4266] [General]
Hii Raylib is interesting project
i am a CSharp developer and a C learner.
i am currently working on making a library for Raylib Support in Microsoft Small Basic.
But i have to use Raylib_cs version 3.1.5 for .Net 4.5 Framework Which use Raylib 3.0
My Question is can i use Raylib_cs latest version in .Net 4.5

**Comment by orcmid** [comment]
Fascinating!!

I had not realized Microsoft had done [Small Basic](https://en.wikipedia.org/wiki/Microsoft_Small_Basic) as a kind of starter text-based language.

## About raylib-cs

The latest release, `raylib-cs` 6.1.1 requires `raylib 5.0`.  This is very specific and involves breaking changes to earlier versions of raylib.  This is also not going to work with the current development toward `raylib 5.5` for the same reasons.

Also, `raylib-cs` now targets .NET 5.0.

Note that `raylib-cs` is not maintained at or by raylib and you should obtain it at the
[Chris Dill](https://github.com/ChrisDill/Raylib-cs?tab=readme-ov-file/) github pages.  That is where you need to find detailed technical information.  However, take this under consideration:

> Raylib-cs is passively maintained. Occasional updates may be released from time to time. Pull requests may be accepted if they don't have a large maintenance burden.

## About Microsoft Small Basic Extensions

The latest Microsoft Small Basic release is version 1.2 of October 2015.  There have been no releases of desktop versions since, and now much of the on-line documentation has been removed.  Some of the remaining materials are identified as no longer maintained.

There are also graphics provisions in Small Basic, such as an extensive
[Graphics Window API](https://smallbasic-publicwebsite.azurewebsites.net/docs/GraphicsWindow).

With regard to using raylib-cs in a Small Basic extension.  There is good news and bad news.  

1. The good news is that `raylib-cs` does provide **static** interfaces, satisfying that Small Basic requirement.
2. The bad news is that raylib-cs defines and relies on a significant variety of (structure) types, as does raylib.  This is *not* accommodated in Small Basic and it will be a significant contortion to work around that.  Comparison of the GraphicsWindow API and the raylib-cs API should make that clear.  In particular, the type `Color` has no equivalent in Small Basic, and the limitation is even more significant for the vector types and probably other types of raylib.

**Reply by zs-3** [reply]
Yes you are write small basic dont support vector and anything it is highly basic with only 14 keywords and very less controls of WPF but small basic support extensions.
i have also created a extension for small basic : https://github.com/zs-3/ZS
so we could trasform .NET functions to Small Basic and in next Release of my extension i am adding functions of raylib.
and there are many extension for small basic like Litdev(Most Powerfull),SmallBasicANN(Neural Network) Etc
Litdev : litdev extension for small basic has make small basic as most powerful basic in world through litdev extension in small basic you could make 3D games and physics engine is also in Litdev extension based on Box2D.
Litdev : https://github.com/litdev1/LitDev

Also the reason behind small basic not being update is Microsoft ended its support for small basic so now small basic is alone.

**Reply by zs-3** [reply]
Some games made with litdev extension and small basic.
https://litdev-uk.itch.io/

**Reply by Peter0x44** [reply]
What is the benefit to use this "Small Basic" over Visual Basic .NET, FreeBASIC, Gambas, etc?

**Reply by zs-3** [reply]
Visual Basic.Net has no completion with small basic visual basic is  father of small basic.
i really  never heard of FreeBASIC and Gambas.
So small basic is just the first step for beginners to Visual Basic.Net and CSharp.
in fact you can directly export small basic code to Visual Basic.Net and if you want to export Small Basic code to CSharp then build it application and use .Net decompiler to decompile the exe to csharp or VB.
Also a developer used small basic source code and made new language by evolving it by name Small Visual Basic or sVB.
sVB : https://github.com/VBAndCs/sVB-Small-Visual-Basic

**Reply by zs-3** [reply]
sVB update regularly and is more powerful than SB

**Reply by orcmid** [reply]
1. I think your questions about `raylib-cs` have been answered.  
2. Crafting an extension to map SmallBasic to raylib 5.0 strikes me as considerable scope creep and the performance will not be good because of all the run-time conversions (strings to enumeration values, etc.).  Of course, you get to choose what itch you want to scratch.  Are you prepared to support it and provide materials for learners at the same level as SB is intended?  Is there a community you have identified for users of raylib-SB?  Also, the `raylib` project does not provide assistance with regard to `raylib-cs`; listing the binding is a service, but the bindings are all found and supported elsewhere (if at all).  Are you prepared for that?  Can you obtain assistance at Litdev?
3. It strikes me that sVB might be a better choice.  Is there some specific reason that you are using SB despite SB lack of support now?
